### PR TITLE
KAFKA-17137[part-5]: Ensure Admin APIs are properly tested

### PR DESCRIPTION
**Pull request by TaiJuWu**
jira: https://issues.apache.org/jira/browse/KAFKA-17137

*More detailed description of your change,
createDelegationToken + maxLifeTimeStamp + DELEGATION_TOKEN_EXPIRY_TIME_MS_CONFIG

*Summary of testing strategy (including rationale)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 1]
```diff
@@ -331,6 +332,25 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
 
+  @ParameterizedTest
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testCreateDelegationTokenWithOptions(quorum: String): Unit = {
+    client = createAdminClient
+    val timeout = 1000
+    val options = new CreateDelegationTokenOptions().maxlifeTimeMs(timeout);
+    val token = client.createDelegationToken(options).delegationToken().get()
+    assertEquals(timeout, token.tokenInfo.maxTimestamp - token.tokenInfo.issueTimestamp)
+  }
```
Comment (frankvicky) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: According to the test case, I think this test should cover all options, but only the `maxlifeTimeMs` option is being tested.
Could you test the rest of them ?

**Review by frankvicky — COMMENTED** [review]
Hi @TaiJuWu 
I have a comment, PTAL

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala (hunk 1 above)
Comment (TaiJuWu) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: Hi @frankvicky , thanks for review.
Updated.

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 2]
```diff
@@ -331,6 +332,28 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
   }
 
+  @ParameterizedTest
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testCreateDelegationTokenWithOptions(quorum: String): Unit = {
+    val owner = new KafkaPrincipal("User", "client2")
+    val renewers = List(new KafkaPrincipal("User", "client")).asJava
+    client = createAdminClient
+    val timeout = 1000
+    val options = new CreateDelegationTokenOptions().maxlifeTimeMs(timeout).owner(owner).renewers(renewers)
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: this is duplicate to #16648. this PR should test "delegation.token.expiry.time.ms" and "delegation.token.max.lifetime.ms". For example, this test case should set negative `maxlifeTimeMs` and larger `maxlifeTimeMs` to make sure `maxTimestamp` and `expiryTimestamp` does not exceed the configured limits.

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu thanks for this patch

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 3]
```diff
@@ -331,6 +332,32 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
+    assertEquals(DelegationTokenManagerConfigs.DELEGATION_TOKEN_MAX_LIFE_TIME_MS_DEFAULT, token.tokenInfo.maxTimestamp - token.tokenInfo.issueTimestamp)
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
+  }
+
+  @ParameterizedTest
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testCreateDelegationTokenWithNegativeTimeout(quorum: String): Unit = {
+    client = createAdminClient
+    val timeout = -1
+    assertTrue(timeout < 0)
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: It seems this assert is redundant, right?

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 4]
```diff
@@ -331,6 +332,32 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
     assertFutureExceptionTypeEquals(results.values.get(emptyResourceNameAcl), classOf[InvalidRequestException])
   }
 
+  @ParameterizedTest
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testCreateDelegationTokenWithLargeTimeout(quorum: String): Unit = {
+    client = createAdminClient
+    val timeout = Long.MaxValue
+    assertTrue(timeout > DelegationTokenManagerConfigs.DELEGATION_TOKEN_EXPIRY_TIME_MS_DEFAULT)
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: It seems this assert is redundant, right?

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu thanks for this patch

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala (hunk 3 above)
Comment (TaiJuWu) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: Thanks for review. Removed.

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala (hunk 4 above)
Comment (TaiJuWu) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: Removed.

**Review by chia7712 — APPROVED** [review]
LGTM

**Review by frankvicky — APPROVED** [review]

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 5]
```diff
@@ -331,6 +332,52 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
+
+    assertEquals(DelegationTokenManagerConfigs.DELEGATION_TOKEN_MAX_LIFE_TIME_MS_DEFAULT, token.tokenInfo.maxTimestamp - token.tokenInfo.issueTimestamp)
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
+
+    TestUtils.waitUntilTrue(() => brokers.forall(server => server.tokenCache.tokens().size() == 1),
+      "Timed out waiting for token to propagate to all servers")
+
+    val ExpiredOptions = new ExpireDelegationTokenOptions().expiryTimePeriodMs(Long.MaxValue)
+    client.expireDelegationToken(token.hmac, ExpiredOptions)
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: we should check the returned value, right? for example:
```scala
    val executionException = new ExpireDelegationTokenOptions().expiryTimePeriodMs(token.tokenInfo().maxTimestamp())
    val result = client.expireDelegationToken(token.hmac, executionException).expiryTimestamp().get()
    assertEquals(token.tokenInfo().maxTimestamp(), result)
```

Comment (TaiJuWu) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: Hi @chia7712, Thanks for review. Update. Just a small question.

In the case, I set `expiryTimePeriodMs ` to Long.MAX to check the large number as edge case instead of `maxTimeStamp`.
WDYT?

Comment (chia7712) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: > In the case, I set expiryTimePeriodMs to Long.MAX to check the large number as edge case instead of maxTimeStamp.

That will cause negative ts. see https://issues.apache.org/jira/browse/KAFKA-17415

Hence, please don't set Long.MAX for now

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 6]
```diff
@@ -331,6 +332,54 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
+    assertEquals(DelegationTokenManagerConfigs.DELEGATION_TOKEN_MAX_LIFE_TIME_MS_DEFAULT, token.tokenInfo.maxTimestamp - token.tokenInfo.issueTimestamp)
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
+
+    TestUtils.waitUntilTrue(() => brokers.forall(server => server.tokenCache.tokens().size() == 1),
+      "Timed out waiting for token to propagate to all servers")
+
+    val expiredOptions = new ExpireDelegationTokenOptions().expiryTimePeriodMs(token.tokenInfo.maxTimestamp)
+    val expiredResult = client.expireDelegationToken(token.hmac, expiredOptions)
+
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: this is duplicate to line#371, right?

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 7]
```diff
@@ -331,6 +332,54 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
+    val createOptions = new CreateDelegationTokenOptions().maxlifeTimeMs(timeout)
+    val token = client.createDelegationToken(createOptions).delegationToken().get()
+
+    assertEquals(DelegationTokenManagerConfigs.DELEGATION_TOKEN_MAX_LIFE_TIME_MS_DEFAULT, token.tokenInfo.maxTimestamp - token.tokenInfo.issueTimestamp)
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
+
+    TestUtils.waitUntilTrue(() => brokers.forall(server => server.tokenCache.tokens().size() == 1),
+      "Timed out waiting for token to propagate to all servers")
+
+    val expiredOptions = new ExpireDelegationTokenOptions().expiryTimePeriodMs(token.tokenInfo.maxTimestamp)
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: All we want to verify is the expired ts is never larger the max timestamp, so we should set `maxTimestamp + 1`, right?

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu thanks for your patch!

**Comment by TaiJuWu** [comment]
> @TaiJuWu thanks for your patch!

Thanks for review, updated.

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 8]
```diff
@@ -337,6 +338,53 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
+    assertEquals(DelegationTokenManagerConfigs.DELEGATION_TOKEN_MAX_LIFE_TIME_MS_DEFAULT, token.tokenInfo.maxTimestamp - token.tokenInfo.issueTimestamp)
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
+
+    TestUtils.waitUntilTrue(() => brokers.forall(server => server.tokenCache.tokens().size() == 1),
+      "Timed out waiting for token to propagate to all servers")
+
+    val expiredOptions = new ExpireDelegationTokenOptions().expiryTimePeriodMs(token.tokenInfo.maxTimestamp + 1)
+    val expiredResult = client.expireDelegationToken(token.hmac, expiredOptions)
+
+    assertTrue(token.tokenInfo.maxTimestamp >= expiredResult.expiryTimestamp.get())
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: it should be equal, right? `assertEquals(token.tokenInfo.maxTimestamp, expiredResult.expiryTimestamp.get())`

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu please fix the tests according to #16858

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 9]
```diff
@@ -337,6 +338,53 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testCreateDelegationTokenWithLargeTimeout(quorum: String): Unit = {
+    client = createAdminClient
+    val timeout = Long.MaxValue
+
+    val options = new CreateDelegationTokenOptions().maxlifeTimeMs(timeout)
+    val token = client.createDelegationToken(options).delegationToken().get()
+
+    assertEquals(DelegationTokenManagerConfigs.DELEGATION_TOKEN_MAX_LIFE_TIME_MS_DEFAULT, token.tokenInfo.maxTimestamp - token.tokenInfo.issueTimestamp)
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
```
Comment (m1a2st) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: nit: the `token.tokenInfo` have been used four times, I consider make it a variable is more better, WDTY
```
val tokenInfo = client.createDelegationToken(options).delegationToken().get().tokenInfo
```

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 10]
```diff
@@ -337,6 +338,53 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testCreateDelegationTokenWithNegativeTimeout(quorum: String): Unit = {
+    client = createAdminClient
+    val timeout = -1
+
+    val options = new CreateDelegationTokenOptions().maxlifeTimeMs(timeout)
+    val token = client.createDelegationToken(options).delegationToken().get()
+
+    assertEquals(DelegationTokenManagerConfigs.DELEGATION_TOKEN_MAX_LIFE_TIME_MS_DEFAULT, token.tokenInfo.maxTimestamp - token.tokenInfo.issueTimestamp)
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
```
Comment (m1a2st) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: ditto

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 11]
```diff
@@ -337,6 +338,53 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testExpiredTimeStampLargerThanMaxLifeStamp(quorum: String): Unit = {
+    client = createAdminClient
+    val timeout = -1
+
+    val createOptions = new CreateDelegationTokenOptions().maxlifeTimeMs(timeout)
+    val token = client.createDelegationToken(createOptions).delegationToken().get()
+
+    assertEquals(DelegationTokenManagerConfigs.DELEGATION_TOKEN_MAX_LIFE_TIME_MS_DEFAULT, token.tokenInfo.maxTimestamp - token.tokenInfo.issueTimestamp)
+    assertTrue(token.tokenInfo.maxTimestamp >= token.tokenInfo.expiryTimestamp)
```
Comment (m1a2st) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: ditto

**Review by m1a2st — COMMENTED** [review]
Hello @TaiJuWu, I left some comments, PTAL

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala (hunk 8 above)
Comment (TaiJuWu) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: Removed. Thanks for review.

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala (hunk 9 above)
Comment (TaiJuWu) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: Nice suggestion, updated, thanks.

**Review by m1a2st — APPROVED** [review]
LGTM

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala [hunk 12]
```diff
@@ -415,6 +415,54 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
...
+    val options = new CreateDelegationTokenOptions().maxlifeTimeMs(timeout)
+    val tokenInfo = client.createDelegationToken(options).delegationToken.get.tokenInfo
+
+    assertEquals(maxLifeTime, tokenInfo.maxTimestamp - tokenInfo.issueTimestamp)
+    assertTrue(tokenInfo.maxTimestamp >= tokenInfo.expiryTimestamp)
+  }
+
+  @ParameterizedTest
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testCreateDelegationTokenWithNegativeTimeout(quorum: String): Unit = {
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: This is covered by `testExpiredTimeStampLargerThanMaxLifeStamp`, right?

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu thanks for your patch

File: core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala (hunk 12 above)
Comment (TaiJuWu) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala]: Yes, you are right. It is the subset, removed.
